File: Java/src/main/java/org/source/searching/LinearSearch.java
package org.source.searching;

import java.util.Arrays;

public class LinearSearch<T extends Comparable<T>> {
    /**
     * Implementation of the Linear Search Algorithm using Generics
     * worst case: O(n)
     *
     * @param array Array where the search must be conducted
     * @param value Key value are searching for
     * @return Index of the key value (if the key value is not in the array, return -1)
     */
    public int find(T[] array, T value) {
        for (int i = 0; i < array.length; i++) {
            if (array[i].compareTo(value) == 0) {
                return i;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        // Example with array of integers
        LinearSearch linearSearch = new LinearSearch();
        Integer[] array = {45, 22, 4, 2, 4, 5, 9, 19199, Integer.MAX_VALUE, 7, 91, 1, 5, 0};
        Integer value = 19199;
        System.out.println("We are searching for " + value + " in " + Arrays.toString(array));
        System.out.println("Index of the value is " + linearSearch.find(array, value));

        // if array does not contain the value
        value = 777;
        System.out.println("We are searching for " + value + " in " + Arrays.toString(array));
        System.out.println("Index of the value is " + linearSearch.find(array, value));
    }
}

File: Python/src/searching/linear_search.py
def linear_search(array, value):
    """
    The linear search algorithm that finds the index of a value in an array
    
    Args:
    - array (list): A sorted list of elements (integers, floats, etc.).
    - value: The value to search for in the array.
    
    Returns:
    - int: The index of the value in the array if found, otherwise -1.
    
    Linear Search Algorithm:
    1. Iterate through each element in the array.
    2. If the current element matches the value, return its index.
    3. If the value is not found, return -1.
    
    Complexity:
    - Time: O(n) where n is the number of elements in the array.
    - Space: O(1) (constant space complexity).
    
    """
    for i in range(len(array)):
        if (array[i] == value):
            return i
    return -1

if __name__ == "__main__":
    # Example with array of integers
    array = [1, 2, 3, 5, 7, 9, 27]
    value = 7

    print(f"We are searching for {value} in {array}")
    print(f"Index of the value is {linear_search(array, value)}")
    
    # if array does not contain the value
    value = 59
    print(f"We are searching for {value} in {array}")
    print(f"Index of the value is {linear_search(array, value)}")
        